tree/20_cfe_basics/flag_lib.cf

```diff
@@ -0,0 +1,175 @@
+      "pass1" expression => "any";
+
+  methods:
+      "${flag_unique.value}" usebundle => read_flags;
+
+  reports:
+      "set"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "is_set|flag_reserved_${cname}";
+
+      "free"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "!is_set.!flag_reserved_${cname}";
+    info::
+      "Flag '${cname}' is set"
+                          comment => "${flag_unique.value}",
+                               if => "is_set";
+
+      "Flag '${cname}' is reserved"
+                          comment => "${flag_unique.value}",
+                               if => "!is_set.flag_reserved_${cname}";
+
+      "Flag '${cname}' is free"
+                          comment => "${flag_unique.value}",
+                               if => "!is_set.!flag_reserved_${cname}";
+}
+
+bundle agent set_flag(name) {
+  vars:
+      "cname" string => canonify("${name}");
+      "time" string => strftime("localtime", "%FT%T%z", "${sys.systime}");
+      "flag" data => '{ "${cname}": { "created": "${time}" } }';
+    pass2::
+      "output" data => mergedata("ncf_flags.data", "flag");
+
+  classes:
+      "pass3" expression => "pass2";
+      "pass2" expression => "pass1";
+      "pass1" expression => "any";
+
+  files:
+    pass3::
+      "${configuration.flag_file}"
+        create => "true",
+        template_method => "mustache",
+        edit_template => "${sys.workdir}/modules/templates/datastate.json.tpl",
+        template_data => "@{set_flag.output}",
+        edit_defaults => no_backup,
+        classes => classes_generic("write_ncf_flag_${cname}");
+
+  methods:
+      "${flag_unique.value}" usebundle => read_flags;
+
+  reports:
+      "ok"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "write_ncf_flag_${cname}_ok";
+      "failed"
+        bundle_return_value_index => "1",
+                          comment => "${flag_unique.value}",
+                               if => "write_ncf_flag_${cname}_reached.!write_ncf_flag_${cname}_ok";
+    info::
+      "Flag '${cname}' (created ${time}) was set"
+                          comment => "${flag_unique.value}",
+                               if => "!write_ncf_flag_${cname}_ok";
+
+    error::
+      "Flag '${cname}' (created ${time}) could not be set"
+                          comment => "${flag_unique.value}",
+                               if => "write_ncf_flag_${cname}_reached.!write_ncf_flag_${cname}_ok";
+}
+
+bundle agent read_flags {
+  vars:
+      "flag_unique.value" string => eval("${flag_unique.value}+1", "math", "infix");
+
+    # Pass3 to avoid reading the json at each pass...
+    pass3.file_exists::
+      "ncf_flags.data" data => readjson("${configuration.flag_file}"),
```

Comment:
could you document somewhere in the file the expected format of flag file ?

Reply:
done